### PR TITLE
Added rocmlir-gen support for rock.attention op

### PR DESCRIPTION
**Pull request by ravil-mobile**
Closes ROCmSoftwarePlatform/rocMLIR-internal#956

Example:

```bash
> rocmlir-gen -seq_len 1024 -num_heads 32 -t f32 --operation attention --arch gfx908:sramecc+:xnack-  | rocmlir-opt
module attributes {mhal.arch = "amdgcn-amd-amdhsa:gfx908:sramecc+:xnack-"} {
  func.func @rock_attention(%arg0: memref<1024x32xf32>, %arg1: memref<32x1024xf32>, %arg2: memref<1024x32xf32>, %arg3: memref<1024x1024xf32>, %arg4: memref<1024x32xf32>) attributes {kernel, mhal.arch = "amdgcn-amd-amdhsa:gfx908:sramecc+:xnack-"} {
    rock.attention(%arg0, %arg1, %arg2, %arg3, %arg4) features =  mfma|dot|atomic_add : memref<1024x32xf32>, memref<32x1024xf32>, memref<1024x32xf32>, memref<1024x1024xf32>, memref<1024x32xf32>
    return
  }
}
```

### CODE REVIEW

**Comment by manupak** [comment]
Some github advice :
as this PR is adding rocmlir-gen kernel gen support, lets title it in that way : i.e.. Added rocmlir-gen support for rock.attention op.
Also you could do "closes : (link to issue)" in the description -- so merging this PR will close the issue automatically.

**Comment by manupak** [comment]
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 1]
```diff
@@ -855,18 +866,38 @@ static void populateDefaults() {
 }
 
 static LogicalResult detectMissingArguments() {
-  const static std::vector<const llvm::cl::opt<int64_t> *> requiredConvArgs = {
+  using RequiredArgsType = std::vector<const llvm::cl::opt<int64_t> *>;
+  static RequiredArgsType requiredConvArgs = {
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: I think we can keep the const?

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 2]
```diff
@@ -2807,10 +2902,11 @@ static ModuleOp generateKernel(MLIRContext *context, GenParams &genParams,
   static rock::Conv2dGenerator conv2dGenerator;
 
   bool isGemm = operation == rock::KernelType::Gemm;
+  bool isAttention = operation == rock::KernelType::Attention;
   auto convConfig = populateConvConfig.getValue();
 
-  if (!convConfig.empty() && isGemm) {
-    llvm::errs() << "Cannot use --conv-config with gemm operations\n";
+  if (!convConfig.empty() && (isGemm || isAttention)) {
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: nit : alternatively, we can add isConv

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 3]
```diff
@@ -2919,7 +3018,7 @@ static ModuleOp generateKernel(MLIRContext *context, GenParams &genParams,
   }
 
   // TODO: Extract isApplicable check to be its own component
-  if (!isGemm &&
+  if (!(isGemm || isAttention) &&
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: could use isConv here..

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 4]
```diff
@@ -1884,6 +1914,68 @@ static func::FuncOp createGpuGemmKernel(ModuleOp module,
...
+    module->setAttr("mhal.arch", archAttr);
+
+  SmallVector<Type, 5> argTypes;
+  getAttentionTypes(params.types, argTypes);
+
+  SmallVector<NamedAttribute, 2> funcAttrs = {
+      builder.getNamedAttr("kernel", builder.getUnitAttr()),
+      builder.getNamedAttr("mhal.arch", archAttr)};
+
+  constexpr StringLiteral kernelName("rocm_attention");
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: nit : lets call it rock_attention as rock is our project name and also being consistent with others.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 5]
```diff
@@ -2881,15 +2977,18 @@ static ModuleOp generateKernel(MLIRContext *context, GenParams &genParams,
...
       for (const auto &arg :
            {filterDataType.getValue(), inputDataType.getValue(),
             outputDataType.getValue()})
         genParams.types.push_back(typeFromString(arg, context));
       genParams.convConfig = std::nullopt;
-      (void)createGpuGemmKernel(module, genParams);
+      if (isGemm) {
+        (void)createGpuGemmKernel(module, genParams);
+      } else {
+        (void)createGpuAttentionKernel(module, genParams);
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: this is not related to a gemm; so might as well as a seperate stub for it.
Im a bit confused why/how filterDataType inputDataType outputDataType are relevant/working here.

**Review by manupak — COMMENTED** [review]
The code broadly looks good.
Few suggestions/nits and a question to understand the related-ness towards the gemm kernel.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 2 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: good idea!

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 4 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: sorry, it was typo. I wanted to call it `rock`. Done

**Comment by manupak** [comment]
As for tests (to keep this bit not broken until we have kernel generation), lets add a small test similiar to : https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/develop/mlir/test/rocmlir-gen/gemm-kernel.mlir

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 5 above)
Comment (krzysz00) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: `filter`, `input`, and `output` are also `a`, `b`, and `c` - they're aliased

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 2 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: done

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 3 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: done

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 1 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: I also prefer `const` but I am struggling to bind it to

https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/636f6e72c5abd272502f8264d87838b5f08badd6/mlir/tools/rocmlir-gen/rocmlir-gen.cpp#L880-L885

It is not allowed to bind `const ref` to `ref`. I though it would be ugly to use `const_cast`; `ternary operator` would result in a lengthy expr.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 5 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: I did something like this

https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/754db02dfa0f55f6d0dc045513c40a0a1af3c9f3/mlir/tools/rocmlir-gen/rocmlir-gen.cpp#L2984-L2990

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 1 above)
Comment (ravil-mobile) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: Did something like that

https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/35480b33dcddceca4cb284bc0ad2d735fa2e28a1/mlir/tools/rocmlir-gen/rocmlir-gen.cpp#L877-L900

**Comment by ravil-mobile** [comment]
> As for tests (to keep this bit not broken until we have kernel generation), lets add a small test similiar to : https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/develop/mlir/test/rocmlir-gen/gemm-kernel.mlir

Thanks for the reference.
Done

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 6]
```diff
@@ -2598,6 +2701,9 @@ static LogicalResult populateHostHarnessLogic(
     case rock::KernelType::Conv2DBwdWeight:
       outIndices.push_back(0);
       break;
+    case rock::KernelType::Attention:
```
Comment (manupak) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp:2704]: Maybe lets fail here to be explicit about this for now with a error message

**Review by manupak — COMMENTED** [review]
Just a single comment.

**Review by manupak — APPROVED** [review]
LGTM!
